File: platforms/ios/devkit/Sources/NativeblocksDevkit/feature/logging/data/LoggingRepositoryImpl.swift
import Foundation
import NativeblocksRuntime

internal class LoggingRepositoryImpl: LoggingRepository {
    private static let TAG = "LoggingRepository"

    private let signalingClient: SocketClient
    private let kitEnvironment: DevKitEnvironment
    private let remoteSession: IRemoteSession

    private var uuid: String {
        remoteSession.getUUID()
    }

    private lazy var projectId: String = {
        JWTUtils.decodedBody(kitEnvironment.apiKey)?["id"] as? String ?? ""
    }()

    init(
        signalingClient: SocketClient,
        kitEnvironment: DevKitEnvironment,
        remoteSession: IRemoteSession
    ) {
        self.signalingClient = signalingClient
        self.kitEnvironment = kitEnvironment
        self.remoteSession = remoteSession
    }

    func sendLog(
        level: LoggerEventLevel,
        event: String,
        message: String,
        parameters: [String: String]
    ) async throws {
        do {
            let logMessage = LogEventMessage(
                clientId: uuid,
                roomId: projectId,
                level: level.rawValue,
                event: event,
                message: message,
                parameters: parameters
            )

            signalingClient.send(message: logMessage)
            DevKitLogger.debug(Self.TAG, "Log sent: event=\(event), level=\(level.rawValue)")
        } catch {
            DevKitLogger.debug(Self.TAG, "Failed to send log message: \(error.localizedDescription)")
            throw error
        }
    }
}

File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/feature/logging/data/LoggingRepositoryImpl.kt
package io.nativeblocks.devkit.feature.logging.data

import io.nativeblocks.devkit.feature.live.data.SocketClient
import io.nativeblocks.devkit.feature.live.data.LogEventMessage
import io.nativeblocks.devkit.feature.logging.domain.repository.LoggingRepository
import io.nativeblocks.devkit.session.DevKitEnvironment
import io.nativeblocks.devkit.session.IRemoteSession
import io.nativeblocks.devkit.util.DevKitLogger
import io.nativeblocks.devkit.util.JWT
import io.nativeblocks.runtime.api.provider.logger.LoggerEventLevel

internal class LoggingRepositoryImpl(
    private val signalingClient: SocketClient,
    private val kitEnvironment: DevKitEnvironment,
    private val remoteSession: IRemoteSession,
) : LoggingRepository {

    companion object {
        private val TAG = LoggingRepositoryImpl::class.java.simpleName
    }

    private val uuid: String
        get() = remoteSession.getUUID()

    private val projectId: String by lazy {
        JWT.decodedBody(kitEnvironment.apiKey)?.getString("id") ?: ""
    }

    override suspend fun sendLog(
        level: LoggerEventLevel,
        event: String,
        message: String,
        parameters: Map<String, String>
    ) {
        try {
            signalingClient.send(
                LogEventMessage(
                    clientId = uuid,
                    roomId = projectId,
                    level = level.name,
                    event = event,
                    message = message,
                    parameters = parameters
                )
            )
            DevKitLogger.d(TAG, "Log sent: event=$event, level=$level")
        } catch (e: Exception) {
            DevKitLogger.e(TAG, "Failed to send log message: ${e.message}", e)
        }
    }
}
